src/canvas-extractor/canvas_extractor/api/authentication_events.py

```diff
@@ -0,0 +1,144 @@
+# SPDX-License-Identifier: Apache-2.0
+# Licensed to the Ed-Fi Alliance under one or more agreements.
+# The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+# See the LICENSE and NOTICES files in the project root for more information.
+
+import logging
+from typing import List
+import re
+
+from pandas import DataFrame
+import sqlalchemy
+from canvasapi.authentication_event import AuthenticationEvent
+from canvasapi.user import User
+from canvasapi.paginated_list import PaginatedList
+
+from .canvas_helper import to_df
+from .resource_sync import cleanup_after_sync, sync_to_db_without_cleanup
+from .api_caller import call_with_retry
+
+AUTH_EVENTS_RESOURCE_NAME = "Authentication_Events"
+
+logger = logging.getLogger(__name__)
+
+
+def custom_get_new_page(self: PaginatedList):
+    response = self._requester.request(
+        self._request_method, self._next_url, **self._next_params
+    )
+    data = response.json()
+    self._next_url = None
+
+    next_link = response.links.get("next")
+    regex = r"{}(.*)".format(re.escape(self._requester.base_url))
+
+    self._next_url = (
+        re.search(regex, next_link["url"]).group(1) if next_link else None
+    )
+
+    self._next_params = {}
+
+    content = []
+
+    if self._root:
+        try:
+            data = data[self._root]
+        except KeyError:
+            # TODO: Fix this message to make more sense to an end user.
+            raise ValueError("Invalid root value specified.")
+
+    if "audit/authentication/users" in self._first_url:
+        data = data['events']
+
+    for element in data:
+        if element is not None:
+            element.update(self._extra_attribs)
+            content.append(self._content_class(self._requester, element))
+
+    return content
+
+
+PaginatedList._get_next_page = custom_get_new_page
```

Comment:
When you do the monkey patch here, does the new method still work for the other API calls, or is there a required ordering for patching?

Reply:
I've often seen a patching pattern where you keep a reference to the original method, swap in the new method so it can be used, then later swap back to the original method.

Reply:
This will apply for all calls, I am not sure if this is the right place to leave this code, but the patch is not invassive, it actually is a copy of the original method with an extra validation for login activities.